// @vitest/coverage-c8@0.29.2/dist/index.js
import inspector from 'inspector';

const session = new inspector.Session();
function startCoverage() {
  session.connect();
  session.post("Profiler.enable");
  session.post("Profiler.startPreciseCoverage", {
    callCount: true,
    detailed: true
  });
}
async function takeCoverage() {
  return new Promise((resolve, reject) => {
    session.post("Profiler.takePreciseCoverage", async (error, coverage) => {
      if (error)
        return reject(error);
      const result = coverage.result.filter(filterResult);
      resolve({ result });
    });
  });
}
function stopCoverage() {
  session.post("Profiler.stopPreciseCoverage");
  session.post("Profiler.disable");
  session.disconnect();
}
function filterResult(coverage) {
  if (!coverage.url.startsWith("file://"))
    return false;
  if (coverage.url.includes("/node_modules/"))
    return false;
  return true;
}

async function getProvider() {
  const { C8CoverageProvider } = await import('./provider-a020853a.js');
  return new C8CoverageProvider();
}

export { getProvider, startCoverage, stopCoverage, takeCoverage };
